benchmark/Benchmark.cpp

```diff
@@ -0,0 +1,125 @@
+    _registerdBenchmarks.push_back(benchmark);
+  }
+}
+
+// ____________________________________________________________________________
+float BenchmarkRecords::measureTimeOfFunction(
+    const std::function<void()>& functionToMeasure) const {
+  ad_utility::Timer benchmarkTimer;
+     
+  benchmarkTimer.start();
+  functionToMeasure();
+  benchmarkTimer.stop();
+
+  return benchmarkTimer.secs();
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addSingleMeasurment(const std::string descriptor,
+      const std::function<void()>& functionToMeasure) {
+    _singleMeasurments.push_back(RecordEntry{descriptor,
+        measureTimeOfFunction(functionToMeasure)});
+}
+
+// ____________________________________________________________________________
+const std::vector<BenchmarkRecords::RecordEntry>&
+   BenchmarkRecords::getSingleMeasurments() const{
+  return _singleMeasurments;
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addGroup(const std::string descriptor) {
+  // Is there already a group with this descriptor? If so, that is not allowed.
+  AD_CHECK(_recordGroups.find(descriptor) == _recordGroups.end());
+
+  // There is no group, so create one without any entries and add them to
+  // the hash map.
+  _recordGroups[descriptor] = BenchmarkRecords::RecordGroup{descriptor, {}};
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addToExistingGroup(const std::string& groupDescriptor,
+    const std::string descriptor,
+    const std::function<void()>& functionToMeasure) {
+  // Does the group exist?
+  auto groupEntry = _recordGroups.find(groupDescriptor);
+  AD_CHECK(groupEntry != _recordGroups.end());
+
+  // Add the descriptor and measured time to the group.
+  groupEntry->second.entries.push_back(
+      RecordEntry{descriptor, measureTimeOfFunction(functionToMeasure)});
+}
+
+// ____________________________________________________________________________
+const ad_utility::HashMap<std::string, BenchmarkRecords::RecordGroup>&
+   BenchmarkRecords::getGroups() const {
+  return _recordGroups;
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addTable(const std::string descriptor,
+    const std::vector<std::string> rowNames,
+    const std::vector<std::string> columnNames) {
+  // Is there already a table with this descriptor? If so, that is not allowed.
+  AD_CHECK(_recordTables.find(descriptor) == _recordTables.end());
+
+  // Add a new entry.
+  _recordTables[descriptor] = BenchmarkRecords::RecordTable{descriptor,
+    rowNames, columnNames, {}};
+
+  // We already know the size of the table, so lets resize the vectors
+  // and give every entry a default value of 0.
+  _recordTables[descriptor].entries.resize(rowNames.size(),
+      std::vector<float>(rowNames.size(), 0));
```

Comment:
This code should probably be in the constructor of the `BenchmarkRecors` class.

Reply:
I had to create two constructors, because without a default constructor, some of the types created through the headers don't work.

See [8d1ea60](https://github.com/ad-freiburg/qlever/pull/860/commits/8d1ea603253611ec75419afc3d35faba781cf778).